### PR TITLE
Fix issue around i64 neg one check in binaryen.js / asmjs

### PR DESCRIPTION
**Pull request by MaxGraey**
Since #2870 I refactored this line: https://github.com/WebAssembly/binaryen/pull/2870/files#diff-54e02686df3480c98e3f90a24705c00bR1395

from 
```cpp
if (right->value == Literal(int32_t(-1)) ||
    right->value == Literal(int64_t(-1))) {
```
to
```cpp
auto constRight = right->value.getInteger();
if (constRight == -1LL) {
```

and it's working for native version as you can see in extra tests here for example:
```wat
(drop (i64.le_u    ;; skip
  (local.get $y)
  (i64.const 4294967295)
))
```
don't optimize to `i64.const 1` but if you build same example via binaryen.js this happening. Accidentally `4294967295` constant in binaryen.js  interprets as `u64(-1)` and test above folded to:
```wat
(drop (i32.const 1))
```
Which is wrong and valid only for `(i32.const 4294967295)` but not for `(i64.const 4294967295)`.

I'm not sure how better fix this. We could revert old approach with checking Literal values. Or may be it's edge case of `i64-to-i32-lowering` or something other which relate to wasm2js and better fix that pass?

cc @kripken

### CODE REVIEW

File: test/passes/optimize-instructions_all-features.txt [hunk 1]
```diff
@@ -3381,6 +3381,18 @@
...
+   (local.get $x)
+  )
+  (drop
+   (local.get $y)
+  )
+  (drop
+   (i64.and
+    (local.get $y)
+    (i64.const 4294967295)
+   )
```
Comment (MaxGraey) [inline test/passes/optimize-instructions_all-features.txt:3394]: So it's right with native binaryen but with binaryen.js this folded to
```wat
(drop (local.get $y))
```

File: test/passes/optimize-instructions_all-features.txt [hunk 2]
```diff
@@ -3422,6 +3434,12 @@
   (drop
    (i32.const 1)
   )
+  (drop
+   (i64.le_u
+    (local.get $y)
+    (i64.const 4294967295)
+   )
+  )
```
Comment (MaxGraey) [inline test/passes/optimize-instructions_all-features.txt:3442]: similar to prev on `binaryen.js` this folded to
```wat
(drop (i32.const 1))
```

So it misinterpretation `i64.const 4294967295` as `i64.const -1`

**Comment by kripken** [comment]
It's good we found this - let's fix the underlying issue though, instead of working around it here. (Do we have current breakage that we need to fix urgently, though? If so we may want to land this, or revert the PR, or something else for now.)

Can you make a standalone testcase in C or Cpp that shows the problem? Then I can debug that.

**Comment by MaxGraey** [comment]
> Can you make a standalone testcase in C or Cpp that shows the problem? Then I can debug that.

Hmm, I will try. My guess it somehow related to process how union with `int64` lowering in asm.js
```cpp
union Integer {
   int32_t i32;
   int_64_t i64;
}

Integer val;
val.i64 = -1LL;
val.i32 = -1;

assert(val.i64 != 4294967295);
```

But I'm not sure

**Comment by MaxGraey** [comment]
No, unfortunately I still not isolate this issue to minimal example.

**Comment by MaxGraey** [comment]
But I can confirm this PR fixed problem, so will be great merge it as fast workaround

**Comment by kripken** [comment]
What's a minimal `.wat` file to show the issue, at least - that would help too? From the description it sounds like

```wat
(drop (i64.le_u    ;; skip
  (local.get $y)
  (i64.const 4294967295)
))
```

Or something like that, but would be good to have a full `.wat` file here, with the expected and incorrect output.

**Comment by MaxGraey** [comment]
> Or something like that, but would be good to have a full .wat file here, with the expected and incorrect output.

yes, It's minimal example for `binaryen.js`. So native binaryen don't tuch this and optimize to
```wat
(drop (i64.le_u
  (local.get $y)
  (i64.const 4294967295)
))
```
but binaryen.js unexpectedly fold it to
```wat
(drop (i32.const 1))
```
Which should valid only for `(i64.const -1)` or `(i32.const -1)` (4294967295)

**Comment by kripken** [comment]
I see, thanks @MaxGraey 

Ok, with that I could reproduce the problem, and then I spent some time reducing it, ending up with this bug report in LLVM: https://bugs.llvm.org/show_bug.cgi?id=46447 - this looks like an LLVM wasm backend codegen issue. Very good we found it!

**Comment by MaxGraey** [comment]
@kripken As I understand this will not be quick fix. First of all it should fixed on LLVM, than Binaryen should switch to this LLVM version. Could we merge this workaround as is but with `// TODO: revert changes after fix https://bugs.llvm.org/show_bug.cgi?id=46447` comment?

**Comment by kripken** [comment]
If there's not a quick fix on the LLVM side then I agree.

But if there is, then waiting for that isn't so bad as emscripten updates to latest LLVM very quickly - we can make sure to tag a release right after.

So it depends on what @tlively says about the LLVM status I think.

**Comment by MaxGraey** [comment]
Ah, good to know binaryen can fast migrate to newer LLVM

**Comment by tlively** [comment]
I have a fix up for review at https://reviews.llvm.org/D83017, so this shouldn’t be a problem much longer.

**Comment by MaxGraey** [comment]
@tlively Thanks!

**Comment by MaxGraey** [comment]
Fix already in [upstream of LLVM](https://github.com/llvm/llvm-project/commit/8df30d988e9e595fa9883706198aec23b2e6d227). Good job @tlively !

**Comment by kripken** [comment]
Does the binaryen.js buildbot use the `tot` build? Then it should already have the fix.

If it uses a release, then I can tag a release after the weekend.

**Comment by dcodeIO** [comment]
Yep, it uses `tot` by default.

**Comment by kripken** [comment]
Ok great, then this PR should not be needed anymore (please confirm).

Thanks again for filing this @MaxGraey - if this LLVM bug wasn't found on binaryen, it might have affected other projects that are much harder (closed source, bigger, etc.) for us to get a useful testcase from!

**Comment by MaxGraey** [comment]
Yes, everything worked out very well. Thank you all for your quick and efficient fix! Will close this PR tomorrow after check new binaryen.js release.

**Comment by MaxGraey** [comment]
@kripken 
> If it uses a release, then I can tag a release after the weekend.

It seems binaryen.js bot required new commmit or release of emscripten. Without this it download this most recent version:
`emcc (Emscripten gcc/clang-like replacement) 1.39.18 (2bee818f21ff1687f28a4186795c10509cea784c)`

**Comment by dcodeIO** [comment]
Binaries the bot used upon `emsdk install tot` last run:

```
https://storage.googleapis.com/webassembly/emscripten-releases-builds/linux/99913ebee94cf357e1df3bbe7e021f83e91ebc5c/wasm-binaries.tbz2
```

**Comment by tlively** [comment]
Our tree is broken for other reasons, so the fix will not roll into tot until tomorrow probably

**Comment by MaxGraey** [comment]
Can confirm LLVM fix works and recent binaryen.js already shipped with proper switch logic. @tlively thanks again! Close this

**Pull request closed without merging**